File: test/jdk/jdk/incubator/concurrent/ScopedValue/Stress.java
/*
 * Copyright (c) 2021, 2022 Red Hat, Inc. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

/**
 * @test
 * @summary Stress test for ScopedValue
 * @modules jdk.incubator.concurrent
 * @compile --enable-preview -source ${jdk.version} Stress.java
 * @run testng/othervm/timeout=300 -XX:-TieredCompilation --enable-preview Stress
 * @run testng/othervm/timeout=300 --enable-preview Stress
 */

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadLocalRandom;
import jdk.incubator.concurrent.ScopedValue;
import jdk.incubator.concurrent.StructureViolationException;
import jdk.incubator.concurrent.StructuredTaskScope;
import org.testng.annotations.Test;
import static org.testng.Assert.*;

public class Stress {
    public static final ScopedValue<Integer> el = ScopedValue.newInstance();

    public static final ScopedValue<Integer> inheritedValue = ScopedValue.newInstance();

    final ThreadLocalRandom tlr = ThreadLocalRandom.current();
    static RuntimeException ex = new RuntimeException();
    int ITERS = 100_000;

    class DeepRecursion implements Runnable {
        public void run() {
            final var last = el.get();
            ITERS--;
            var nextRandomFloat = tlr.nextFloat();
            try {
                ScopedValue.where(el, el.get() + 1).run(() -> fibonacci_pad(20, this));
                if (!last.equals(el.get())) {
                    throw ex;
                }
            } catch (StackOverflowError e) {
                if (nextRandomFloat <= 0.1) {
                    ScopedValue.where(el, el.get() + 1).run(this);
                }
            } catch (StructureViolationException structureViolationException) {
                // Can happen if the stack overflow prevented a StackableScope from
                // being removed. We can continue.
            } finally {
                if (!last.equals(el.get())) {
                    throw ex;
                }
            }

            Thread.yield();
        }
    }

    static final Runnable nop = new Runnable() {
        public void run() { }
    };

    long fibonacci_pad1(int n, Runnable op) {
        if (n <= 1) {
            op.run();
            return n;
        }
        return fibonacci_pad1(n - 1, op) + fibonacci_pad1(n - 2, nop);
    }

    private static final Integer I_42 = 42;

    long fibonacci_pad(int n, Runnable op) {
        final var last = el.get();
        try {
            return fibonacci_pad1(tlr.nextInt(n), op);
        } catch (StackOverflowError err) {
            if (!inheritedValue.get().equals(I_42)) {
                throw ex;
            }
            if (!last.equals(el.get())) {
                throw ex;
            }
            throw err;
        }
    }

    void runInNewThread(Runnable op) {
        try (var scope = new StructuredTaskScope<Object>()) {
            var future = scope.fork(() -> {
                op.run();
                return null;
            });
            future.get();
            scope.join();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void run() {
        try {
            ScopedValue.where(inheritedValue, 42).where(el, 0).run(() -> {
                try (var scope = new StructuredTaskScope<Object>()) {
                    try {
                        if (tlr.nextBoolean()) {
                            final var deepRecursion = new DeepRecursion();
                            deepRecursion.run();
                        } else {
                            Runnable op = new Runnable() {
                                public void run() {
                                    try {
                                        fibonacci_pad(20, this);
                                    } catch (StackOverflowError e) {
                                        if (!inheritedValue.get().equals(I_42)) {
                                            throw ex;
                                        }
                                    }
                                }
                            };
                            runInNewThread(op);
                        }
                        scope.join();
                    } catch (StructureViolationException structureViolationException) {
                        // Can happen if a stack overflow prevented a StackableScope from
                        // being removed. We can continue.
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                }
            });
        } catch (StructureViolationException structureViolationException) {
            // Can happen if a stack overflow prevented a StackableScope from
            // being removed. We can continue.
        }
    }

    @Test
    public void doTest() {
        while (ITERS > 0) {
            run();
        }
        System.out.println("OK");
    }
}

File: src/hotspot/share/compiler/stress.cpp
/*
 * Copyright (c) 2026, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 *
 */

#include "compiler/compileLog.hpp"
#include "compiler/compiler_globals.hpp"
#include "compiler/compilerDefinitions.hpp"
#include "compiler/compilerDirectives.hpp"
#include "compiler/stress.hpp"
#include "runtime/globals_extension.hpp"
#include "runtime/os.hpp"
#include "utilities/ticks.hpp"

#ifdef COMPILER2
#include "opto/c2_globals.hpp"
#endif // COMPILER

static bool should_initialize_stress_seed(CompilerType comp) {
  switch (comp) {
#ifdef COMPILER1
    case compiler_c1:
      return false;
#endif // COMPILER1
#ifdef COMPILER2
    case compiler_c2:
      return StressLCM || StressGCM || StressIGVN || StressCCP ||
        StressIncrementalInlining || StressMacroExpansion ||
        StressMacroElimination || StressUnstableIfTraps ||
        StressBailout || StressLoopPeeling || StressCountedLoop ||
        StressEliminateAllocations;
#endif // COMPILER2
    default:
      assert(comp != compiler_none && comp != compiler_number_of_types, "expected valid compiler");
      return false;
  }
}

Stress::Stress(DirectiveSet* directives, CompileLog* log, CompilerType comp) {
  if (!should_initialize_stress_seed(comp)) {
    _stress_seed = 0;
    return;
  }

  if (FLAG_IS_DEFAULT(StressSeed) || (FLAG_IS_ERGO(StressSeed) && directives->RepeatCompilationOption)) {
    _stress_seed = static_cast<uint>(Ticks::now().nanoseconds());
    FLAG_SET_ERGO(StressSeed, _stress_seed);
  } else {
    _stress_seed = StressSeed;
  }
  if (log != nullptr) {
    log->elem("stress_test seed='%u'", _stress_seed);
  }
}

uint Stress::random() {
  _stress_seed = os::next_random(_stress_seed);
  return _stress_seed;
}

const uint RANDOMIZED_DOMAIN_POW  = 29;
const uint RANDOMIZED_DOMAIN      = 1 << RANDOMIZED_DOMAIN_POW;
const uint RANDOMIZED_DOMAIN_MASK = (1 << (RANDOMIZED_DOMAIN_POW + 1)) - 1;

// This method can be called an arbitrary number of times, with the current count
// as the argument. The logic allows for selecting a single candidate from the
// running list of candidates as follows:
//    int count = 0;
//    Cand* selected = null;
//    while(cand = cand->next()) {
//      if (randomized_select(++count)) {
//        selected = cand;
//      }
//    }
//
// Including the count equalizes the chances any candidate is "selected".
// This is useful when we don't have the complete list of candidates to choose
// from uniformly. In this case, we need to adjust the randomicity of the
// selection, or else we will end up biasing the selection towards the latter
// candidates.
//
// A quick back-of-the-envelope calculation shows that for the list of n candidates
// the equal probability for the candidate to persist as "best" as can be
// achieved by replacing it with "next" k-th candidate with the probability
// of 1/k. It can be easily shown that by the end of the run, the
// probability for any candidate has converged to 1/n, thus giving the
// uniform distribution among all the candidates.
//
// We don't care about the domain size as long as (RANDOMIZED_DOMAIN / count) is large.
bool Stress::randomized_select(uint count) {
  return (random() & RANDOMIZED_DOMAIN_MASK) < (RANDOMIZED_DOMAIN / count);
}
